### PR TITLE
Tests for cache system

### PR DESCRIPTION
**Pull request by Mrowqa**
Based on #223. Draft to avoid merging before #223.

### CODE REVIEW

File: wasmtime-environ/src/cache.rs [hunk 1]
```diff
@@ -505,3 +505,343 @@ impl<'de> Deserialize<'de> for JtOffsetsWrapper<'_> {
...
+        ))
+    }
+
+    impl ModuleCacheEntry {
+        pub fn mod_cache_path(&self) -> &Option<PathBuf> {
+            &self.mod_cache_path
+        }
+    }
+
+    // cranelift's types (including SecondaryMap) doesn't implement PartialEq
```
Comment (sunfishcode) [inline wasmtime-environ/src/cache.rs]: I know I originally suggested using this approach, but I actually think we can go back and implement `Eq`/`PartialEq` for `SecondaryMap`, with the approach of trimming the trailing default-value elements. That should hopefully allow the code here to be much simpler.

File: wasmtime-environ/src/cache.rs [hunk 2]
```diff
@@ -505,3 +505,343 @@ impl<'de> Deserialize<'de> for JtOffsetsWrapper<'_> {
...
 }
+
+//-////////////////////////////////////////////////////////////////////
+//-//////////////////////////// Tests /////////////////////////////////
+//-////////////////////////////////////////////////////////////////////
+
+// Since cache system is a global thing, each test needs to be run in seperate process.
+// So, init() tests are run as integration tests.
+// However, caching is a private thing and needs to be tested internally. Thus we have
+// one big test here.
```
Comment (sunfishcode) [inline wasmtime-environ/src/cache.rs]: This is a pretty big test; would it make sense to move it to a separate file in the `wasmtime-environ/tests` directory? Or is this comment saying that it's using things that aren't visible outside this file?

File: wasmtime-environ/src/cache.rs [hunk 3]
```diff
@@ -505,3 +505,343 @@ impl<'de> Deserialize<'de> for JtOffsetsWrapper<'_> {
...
+                if it_lhs.zip(it_rhs).any(|(lhs, rhs)| lhs != rhs) {
+                    return false;
+                }
+            }
+
+            true
+        }
+    }
+
+    // binemit::Reloc doesn't implement PartialEq
```
Comment (sunfishcode) [inline wasmtime-environ/src/cache.rs]: We should make `binemit::Reloc` implement `PartialEq` and `Eq`.

File: wasmtime-environ/src/cache.rs (hunk 2 above)
Comment (Mrowqa) [reply wasmtime-environ/src/cache.rs]: We can't load nor store cache outside the `wasmtime-environ`. The idea is that it's private, an implementation detail. The `wasmtime-environ/tests` directory is for integration tests, so we can't access private items. Thus, we have to have a tests submodule in the cache module. If you want to, I can move it to `wasmtime-environ/src/cache/tests.rs`.

File: wasmtime-environ/src/cache.rs (hunk 1 above)
Comment (Mrowqa) [reply wasmtime-environ/src/cache.rs]: As we discussed, I'll do it as a follow up refactor later, also implementing the serialization for `SecondaryMap`.

File: wasmtime-environ/src/cache.rs (hunk 3 above)
Comment (Mrowqa) [reply wasmtime-environ/src/cache.rs]: Same thing as with `SecondaryMap` - I'll do it later as follow up refactor.

File: wasmtime-environ/src/cache.rs (hunk 2 above)
Comment (sunfishcode) [reply wasmtime-environ/src/cache.rs]: Yes, moving it to wasmtime-environ/src/cache/tests.rs sounds good.

**Pull request merged**